edb/pgsql/resolver/resolver.py

```diff
@@ -0,0 +1,645 @@
+        result.columns = [
+            context.Column(name=col.name, reference_as=col.name)
+            for col in subctx.scope.rel.columns
+        ]
+
+        ctx.scope.rel = result
+
+    return pgast.RangeSubselect(
+        subquery=subquery,
+        alias=alias,
+    )
+
+
+@dispatch.resolve_range_var.register
+def resolve_RelRangeVar(
+    range_var: pgast.RelRangeVar,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.RelRangeVar:
+    ctx.include_inherited = range_var.include_inherited
+
+    return pgast.RelRangeVar(
+        relation=dispatch.resolve(range_var.relation, ctx=ctx),
+        alias=alias,
+    )
+
+
+@dispatch.resolve_range_var.register
+def resolve_RangeFunction(
+    range_var: pgast.RangeFunction,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.RangeFunction:
+    raise errors.UnsupportedFeatureError('range functions are not supported')
+
+
+@dispatch.resolve_range_var.register
+def resolve_JoinExpr(
+    range_var: pgast.JoinExpr,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.JoinExpr:
+
+    with ctx.empty() as subctx:
+        larg = dispatch.resolve(range_var.larg, ctx=subctx)
+        l_rel = subctx.scope.rel
+
+    with ctx.empty() as subctx:
+        rarg = dispatch.resolve(range_var.rarg, ctx=subctx)
+        r_rel = subctx.scope.rel
+
+    ctx.scope.join_relations.extend((l_rel, r_rel))
+    ctx.scope.tables.extend((l_rel, r_rel))
+
+    quals: Optional[pgast.BaseExpr] = None
+    if range_var.quals:
+        quals = dispatch.resolve(range_var.quals, ctx=ctx)
+
+    if range_var.using_clause:
+        for c in range_var.using_clause:
+            with ctx.empty() as subctx:
+                subctx.scope.tables = [l_rel]
+                l_expr = dispatch.resolve(c, ctx=subctx)
+            with ctx.empty() as subctx:
+                subctx.scope.tables = [r_rel]
+                r_expr = dispatch.resolve(c, ctx=subctx)
+            quals = pgastutils.extend_binop(
+                quals,
+                pgast.Expr(
+                    kind=pgast.ExprKind.OP,
+                    name='=',
+                    lexpr=l_expr,
+                    rexpr=r_expr,
+                ),
+            )
+
+    return pgast.JoinExpr(
+        type=range_var.type,
+        larg=larg,
+        rarg=rarg,
+        quals=quals,
+        alias=alias,
+    )
+
+
+@dispatch._resolve.register
+def resolve_Relation(
+    relation: pgast.Relation, *, ctx: Context
+) -> pgast.Relation:
+    assert relation.name
+
+    # try a CTE
+    cte = next((t for t in ctx.scope.ctes if t.name == relation.name), None)
+    if cte:
+        ctx.scope.rel.name = cte.name
+        ctx.scope.rel.columns = cte.columns.copy()
+        return pgast.Relation(name=cte.name, schemaname=None)
+
+    # lookup the object in schema
+    obj: Optional[s_objtypes.ObjectType] = None
+    if (relation.schemaname or 'public') == 'public':
+        object_type_name = relation.name[0].upper() + relation.name[1:]
```

Comment:
What's going on here?

Reply:
Ok, I've implemented case insensitive lookup we talked about, but there is still a possible problem.

Our ObjectTypes are now exposed as tables with their name converted to lower case, because we want to refer to tables with ident without quotes. But if you have two different ObjectTypes with same name except for casing (`default::User` and `default::user`) they will be exposed as two tables with the same name.